File: src/2022-job-autumn/342.power-of-four.js
/*
 * @lc app=leetcode id=342 lang=javascript
 *
 * [342] Power of Four
 */

// @lc code=start
/**
 * @param {number} n
 * @return {boolean}
 */
var isPowerOfFour = function (n) {
  if (n === 0) {
    return false;
  }

  while (n % 4 === 0) {
    n /= 4;
  }

  return n === 1;
};
// @lc code=end

File: src/2025/342.power-of-four.cpp
/*
 * @lc app=leetcode id=342 lang=cpp
 *
 * [342] Power of Four
 */

#include <cinttypes>

// @lc code=start
class Solution {
 public:
  constexpr bool isPowerOfFour(const int n) {
    if (!IsPowerOfTwo(n)) {
      return false;
    }
    return (n & mask_) == 0;
  }

 private:
  static constexpr std::uint32_t mask_ = 0xAAAAAAAA;
  static constexpr bool IsPowerOfTwo(const int n) {
    return n > 0 && (n & (n - 1)) == 0;
  }
};
// @lc code=end
